testcontainers/src/watchdog.rs

```diff
@@ -0,0 +1,55 @@
+use lazy_static::lazy_static;
+use signal_hook::{
+    consts::{SIGINT, SIGQUIT, SIGTERM},
+    iterator::Signals,
+};
+use std::{collections::BTreeSet, sync::Mutex, thread};
+
+lazy_static! {
+    static ref WATCHDOG: Mutex<Watchdog> =
+        Mutex::new(Watchdog::new().expect("failed to register signal handler"));
+}
+
+use crate::{clients::Cli, core::Docker};
+pub(crate) struct Watchdog {
+    containers: BTreeSet<String>,
+}
+
+impl Watchdog {
+    pub(crate) fn new() -> Result<Self, std::io::Error> {
+        let mut signals = Signals::new(&[SIGTERM, SIGINT, SIGQUIT])?;
+        thread::spawn(move || {
+            let signal_docker = Cli::default();
+            for signal in &mut signals {
+                for container_id in WATCHDOG
+                    .lock()
+                    .map(|s| s.containers.clone())
+                    .unwrap_or_default()
+                {
+                    signal_docker.stop(&container_id);
+                    signal_docker.rm(&container_id);
+                }
+                let _ = signal_hook::low_level::emulate_default_handler(signal);
+            }
+        });
```

Comment:
Using `conquer_once::Lazy` would allow us to spawn this outside of the constructor.

Reply:
Good call :) updated it and pulled it into the lazy function!